sno/working_copy/postgis_adapter.py

```diff
@@ -0,0 +1,280 @@
+        for col in schema
+    ]
+
+    if schema.pk_columns:
+        pk_col_names = (Identifier(col.name) for col in schema.pk_columns)
+        result.append(SQL("PRIMARY KEY({})").format(SQL(", ").join(pk_col_names)))
+
+    return SQL(", ").join(result)
+
+
+def _v2_type_to_pg_type(column_schema, v2_obj):
+    """Convert a v2 schema type to a postgis type."""
+
+    v2_type = column_schema.data_type
+    extra_type_info = column_schema.extra_type_info
+
+    pg_type_info = _V2_TYPE_TO_PG_TYPE.get(v2_type)
+    if pg_type_info is None:
+        raise ValueError(f"Unrecognised data type: {v2_type}")
+
+    if isinstance(pg_type_info, dict):
+        return pg_type_info.get(extra_type_info.get("size", 0))
+
+    pg_type = pg_type_info
+    if pg_type == "geometry":
+        geometry_type = extra_type_info.get("geometryType")
+        crs_name = extra_type_info.get("geometryCRS")
+        crs_id = None
+        if crs_name is not None:
+            crs_id = crs_util.get_identifier_int_from_dataset(v2_obj, crs_name)
+        return _v2_geometry_type_to_pg_type(geometry_type, crs_id)
+
+    if pg_type == "text":
+        length = extra_type_info.get("length", None)
+        return f"varchar({length})" if length is not None else "text"
+
+    if pg_type == "numeric":
+        precision = extra_type_info.get("precision", None)
+        scale = extra_type_info.get("scale", None)
+        if precision is not None and scale is not None:
+            return f"numeric({precision},{scale})"
+        elif precision is not None:
+            return f"numeric({precision})"
+        else:
+            return "numeric"
+
+    return pg_type
+
+
+def _v2_geometry_type_to_pg_type(geometry_type, crs_id):
+    if geometry_type is not None:
+        geometry_type = geometry_type.replace(" ", "")
+
+    if geometry_type is not None and crs_id is not None:
+        return f"geometry({geometry_type},{crs_id})"
+    elif geometry_type is not None:
+        return f"geometry({geometry_type})"
+    else:
+        return "geometry"
+
+
+def postgis_to_v2_schema(
+    pg_table_info, pg_geometry_columns, pg_spatial_ref_sys, id_salt
+):
+    """Generate a V2 schema from the given postgis metadata tables."""
+    return Schema(
+        [
+            _postgis_to_column_schema(
+                col, pg_geometry_columns, pg_spatial_ref_sys, id_salt
+            )
+            for col in pg_table_info
+        ]
+    )
+
+
+def _postgis_to_column_schema(
+    pg_col_info, pg_geometry_columns, pg_spatial_ref_sys, id_salt
+):
+    """
+    Given the postgis column info for a particular column, and some extra context in
+    case it is a geometry column, converts it to a ColumnSchema. The extra context will
+    only be used if the given pg_col_info is the geometry column.
+    Parameters:
+    pg_col_info - a single column from pg_table_info.
+    pg_geometry_columns - contents of the "geometry_columns" table.
+    pg_spatial_ref_sys - contents of the "spatial_ref_sys" table.
```

Comment:
this normally has thousands of rows in it... do we need to extract/iterate through it vs doing a `SELECT...WHERE`?

Reply:
similarly (though less likely) we have some databases with tens of thousands of rows in `geometry_columns` too...

Reply:
The docstring was actually wrong - we are doing a `SELECT...WHERE` and these only contained the geometry columns or spatial reference systems for the given dataset.

Slightly modified - pg_geometry_columns is no longer passed as an extra table, but is instead joined to the pg_col_info as a couple of extra fields.